lib/puppet/provider/cisco_portchannel_global/nxapi.rb

```diff
@@ -0,0 +1,256 @@
+  desc 'The new NXAPI provider.'
+
+  confine feature: :cisco_node_utils
+  defaultfor operatingsystem: :nexus
+
+  mk_resource_methods
+
+  PC_GLOBAL_NON_BOOL_PROPS = [
+    :bundle_hash,
+    :bundle_select,
+    :hash_distribution,
+    :hash_poly,
+    :load_defer,
+    :rotate,
+  ]
+  PC_GLOBAL_BOOL_PROPS = [
+    :asymmetric,
+    :concatenation,
+    :resilient,
+    :symmetry,
+  ]
+  PC_GLOBAL_ALL_PROPS = PC_GLOBAL_NON_BOOL_PROPS + PC_GLOBAL_BOOL_PROPS
+
+  PuppetX::Cisco::AutoGen.mk_puppet_methods(:non_bool, self, '@pc_global',
+                                            PC_GLOBAL_NON_BOOL_PROPS)
+  PuppetX::Cisco::AutoGen.mk_puppet_methods(:bool, self, '@pc_global',
+                                            PC_GLOBAL_BOOL_PROPS)
+
+  def initialize(value={})
+    super(value)
+    @pc_global = Cisco::PortChannelGlobal.globals[@property_hash[:name]]
+    @property_flush = {}
+    debug 'Created provider instance of cisco_portchannel_global'
+  end
+
+  def self.properties_get(global_id, v)
+    debug "Checking instance, global #{global_id}"
+    current_state = {
+      name: global_id
+    }
+
+    # Call node_utils getter for each property
+    PC_GLOBAL_NON_BOOL_PROPS.each do |prop|
+      current_state[prop] = v.send(prop)
+    end
+
+    PC_GLOBAL_BOOL_PROPS.each do |prop|
+      val = v.send(prop)
+      if val.nil?
+        current_state[prop] = nil
+      else
+        current_state[prop] = val ? :true : :false
+      end
+    end
+    new(current_state)
+  end # self.properties_get
+
+  def self.instances
+    globals = []
+    Cisco::PortChannelGlobal.globals.each do |global_id, v|
+      globals << properties_get(global_id, v)
+    end
+    globals
+  end
+
+  def self.prefetch(resources)
+    globals = instances
+
+    resources.keys.each do |id|
+      provider = globals.find { |global| global.instance_name == id }
+      resources[id].provider = provider unless provider.nil?
+    end
+  end # self.prefetch
+
+  def instance_name
+    name
+  end
+
+  def properties_set
+    PC_GLOBAL_ALL_PROPS.each do |prop|
+      next unless @resource[prop]
+      unless @property_flush[prop].nil?
+        @pc_global.send("#{prop}=", @property_flush[prop]) if
+          @pc_global.respond_to?("#{prop}=")
+      end
+    end
+    # custom setters which require one-shot multi-param setters
+    port_channel_load_balance_sym_concat_rot_set
+    port_channel_load_balance_hash_poly_set
+    port_channel_load_balance_asym_rot_set
+  end
+
+  # for n9k/n3k:
+  # port-channel load-balance src-dst ip rotate 4 concatenation symmetric
+  # all the above properties will be set all at once so make sure
+  # all the needed resources are present
+  def port_channel_load_balance_sym_concat_rot_all?
+    @resource[:bundle_hash] &&
+      @resource[:bundle_select] &&
+      @resource[:concatenation] &&
+      @resource[:symmetry] &&
+      @resource[:rotate]
+  end
+
+  def port_channel_load_balance_sym_concat_rot_set
```

Comment:
Ok, after our sync-up I was able to refactor this method to simplify things a bit. I ran the beaker tests following the change and everything still works.  `def port_channel_load_balance_asym_rot_all?` can also be refactored to use this new `flush_boolean?` method.


```
  # We need special handeling for boolean properties in our custom
  # setters below. This helper method returns true if the property
  # flush contains TrueClass or FalseClass values.
  def flush_boolean?(prop)
    @property_flush[prop].is_a?(TrueClass) ||
    @property_flush[prop].is_a?(FalseClass)
  end

  def port_channel_load_balance_sym_concat_rot_set
    return unless port_channel_load_balance_sym_concat_rot_all?

    if @property_flush[:bundle_hash]
      bh = @property_flush[:bundle_hash]
    else
      bh = @pc_global.bundle_hash
    end

    if @property_flush[:bundle_select]
      bs = @property_flush[:bundle_select]
    else
      bs = @pc_global.bundle_select
    end

    if flush_boolean?(:concatenation)
      cc = @property_flush[:concatenation]
    else
      cc = @pc_global.concatenation
    end

    if flush_boolean?(:symmetry)
      sy = @property_flush[:symmetry]
    else
      sy = @pc_global.symmetry
    end

    if @property_flush[:rotate]
      ro = @property_flush[:rotate]
    else
      ro = @pc_global.rotate
    end

    # Additional property dependency checks
    if ro > 0 && cc == false
      fail 'concatenation MUST be true when rotate is non-zero'
    end

    if bs.to_s != 'src-dst' && sy == true
      fail 'Symmetric can be true only for src-dst bundle-select'
    end

    @pc_global.send(:port_channel_load_balance=,
                    bs.to_s, bh.to_s, nil, nil, sy, cc, ro)
  end
```